### PR TITLE
feat: ASR and TTS integration

### PR DESCRIPTION
**Pull request by IgnatovFedor**

### CODE REVIEW

File: deeppavlov/models/nemo/common.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        The same structure where all strings are converted into the base64-encoded bytes wrapped in Binary I/O object.
+
+    """
+    if isinstance(batch, str):
+        return BytesIO(base64.decodebytes(batch.encode()))
+
+    return list(map(ascii_to_bytes_io, batch))
+
+
+@register('bytesIO_to_ascii')
```
Comment (yoptar) [inline deeppavlov/models/nemo/common.py]: I think we should mention `base64` in the name somewhere. Maybe `base64_encode_bitesIO`?

File: deeppavlov/models/nemo/tts.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+import torch
+from nemo.collections.asr.parts import collections, parsers
+from nemo.collections.asr.parts.dataset import TranscriptDataset
+from nemo.collections.tts import TextEmbedding, Tacotron2Encoder, Tacotron2DecoderInfer, Tacotron2Postnet
+from nemo.core.neural_types import NeuralType, LabelsType, LengthsType
+from nemo.utils.decorators import add_port_docs
+from nemo.utils.misc import pad_to
+from scipy.io import wavfile
+from torch import Tensor
+from torch.utils.data import Dataset
```
Comment (kudep) [inline deeppavlov/models/nemo/tts.py:30]: This class isn't used after an import.

**Comment by kudep** [comment]
Please, consider  #1174 PR

**Comment by kudep** [comment]
The rest of PR content is correct as I can see.

File: deeppavlov/models/nemo/tts.py (hunk 2 above)
Comment (kudep) [reply deeppavlov/models/nemo/tts.py:30]: Dataset

File: deeppavlov/models/nemo/asr.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+log = logging.getLogger(__name__)
+
+
+class AudioInferDataset(Dataset):
+    def __init__(self, audio_batch: List[Union[str, BytesIO]], sample_rate: int, int_values: bool, trim=False) -> None:
+        """Dataset reader for AudioInferDataLayer.
+
+        Args:
+            audio_batch: Batch to be red. Elements could be either paths to audio files or Binary I/O objects.
```
Comment (oserikov) [inline deeppavlov/models/nemo/asr.py:40]: red -> read probably?

File: deeppavlov/models/nemo/asr.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            sample_rate: Audio files sample rate.
+            int_values: If true, load samples as 32-bit integers.
+            trim: Trim leading and trailing silence from an audio signal if True.
```
Comment (oserikov) [inline deeppavlov/models/nemo/asr.py:43]: trim_silence probably?

File: deeppavlov/models/nemo/asr.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+                 audio_batch: List[Union[str, BytesIO]],
+                 batch_size: int = 32,
+                 sample_rate: int = 16000,
+                 int_values: bool = False,
+                 trim_silence: bool = False,
+                 **kwargs) -> None:
+        """Initializes Data Loader.
+
+        Args:
+            audio_batch: Batch to be red. Elements could be either paths to audio files or Binary I/O objects.
```
Comment (oserikov) [inline deeppavlov/models/nemo/asr.py:92]: red -> read probably?

File: deeppavlov/models/nemo/asr.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        """
+        data_layer = AudioInferDataLayer(audio_batch=audio_batch, **self.nemo_params['AudioToTextDataLayer'])
+        audio_signal, audio_signal_len = data_layer()
+        processed_signal, processed_signal_len = self.data_preprocessor(input_signal=audio_signal,
+                                                                        length=audio_signal_len)
+        encoded, encoded_len = self.jasper_encoder(audio_signal=processed_signal, length=processed_signal_len)
+        log_probs = self.jasper_decoder(encoder_output=encoded)
+        predictions = self.greedy_decoder(log_probs=log_probs)
+        eval_tensors = [predictions]
+        tensors = self.neural_factory.infer(tensors=eval_tensors)
```
Comment (oserikov) [inline deeppavlov/models/nemo/asr.py:190]: Is `tensors` a good namd for a variable ?

File: deeppavlov/models/nemo/asr.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        text_batch = post_process_predictions(tensors[0], self.labels)
```
Comment (oserikov) [inline deeppavlov/models/nemo/asr.py:191]: Recognized or transcribed text batch provably?

File: deeppavlov/models/nemo/common.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+@register('base64_encode_bytesIO')
+def ascii_to_bytes_io(batch: Union[str, list]) -> Union[BytesIO, list]:
+    """Recursively searches for strings in the input batch and converts them into the base64-encoded bytes wrapped in
+    Binary I/O objects.
+
+    Args:
+        batch: A string or an iterable container with strings at some level of nesting.
+
+    Returns:
+        The same structure where all strings are converted into the base64-encoded bytes wrapped in Binary I/O object.
```
Comment (oserikov) [inline deeppavlov/models/nemo/common.py:44]: ...strings wrapped in .. object**S**?

File: deeppavlov/models/nemo/common.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+    """
+    if isinstance(batch, BytesIO):
+        return base64.encodebytes(batch.read()).decode('ascii')
+
+    return list(map(bytes_io_to_ascii, batch))
+
+
+class NeMoBase(Component, Serializable):
+    """Base class for NeMo Chainer's pipe components."""
```
Comment (oserikov) [inline deeppavlov/models/nemo/common.py:71]: Pipe components or pipeline components?

File: deeppavlov/models/nemo/common.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    def __init__(self, load_path: Union[str, Path], nemo_params_path: Union[str, Path], **kwargs) -> None:
+        """Initializes NeuralModuleFactory on CPU or GPU and reads nemo modules params from yaml.
+
+        Args:
+            load_path: Path to a directory with pretrained checkpoints for NeMo modules.
+            nemo_params_path: Path to a file containig NeMo modules params.
+
+        """
+        super(NeMoBase, self).__init__(save_path=None, load_path=load_path, **kwargs)
+        placement = nemo.core.DeviceType.GPU if torch.cuda.is_available() else nemo.core.DeviceType.CPU
```
Comment (oserikov) [inline deeppavlov/models/nemo/common.py:82]: Iirc this is usually called `device` or smth like that. device_type probably

File: deeppavlov/models/nemo/common.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        self.neural_factory = nemo.core.NeuralModuleFactory(placement=placement)
+        self.modules_to_restore = []
+        self.nemo_params = read_yaml(expand_path(nemo_params_path))
+
+    def __call__(self, *args, **kwargs):
+        raise NotImplementedError
+
+    def load(self) -> None:
+        """Loads prettrained checkpoints for modules from self.modules_to_restore list."""
```
Comment (oserikov) [inline deeppavlov/models/nemo/common.py:91]: pret**~t~**rained

File: deeppavlov/models/nemo/common.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+        module_names = [str(module) for module in self.modules_to_restore]
+        checkpoints = nemo.utils.get_checkpoint_from_dir(module_names, self.load_path)
+        for module, checkpoint in zip(self.modules_to_restore, checkpoints):
+            log.info(f'Restoring {module} from {checkpoint}')
+            module.restore_from(checkpoint)
+
+    def save(self, *args, **kwargs) -> None:
```
Comment (oserikov) [inline deeppavlov/models/nemo/common.py:98]: Is this necessary?

File: deeppavlov/models/nemo/tts.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+        """
+        texts_list, texts_len = zip(*batch)
+        max_len = max(texts_len)
+        max_len = pad_to(max_len, 8)
+
+        texts = torch.empty(len(texts_list), max_len, dtype=torch.long)
+        texts.fill_(pad_id)
+
+        for i, s in enumerate(texts_list):
```
Comment (oserikov) [inline deeppavlov/models/nemo/tts.py:124]: s is probably not unambigous here

File: deeppavlov/models/nemo/tts.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+        """
+        if path_batch is None:
+            path_batch = [BytesIO() for _ in text_batch]
+        elif len(text_batch) != len(path_batch):
+            raise ValueError('Text batch length differs from path batch length.')
+        else:
+            path_batch = [expand_path(path) for path in path_batch]
+
+        data_layer = TextDataLayer(text_batch=text_batch, **self.nemo_params['TranscriptDataLayer'])
+        transcript, transcript_len = data_layer()
```
Comment (oserikov) [inline deeppavlov/models/nemo/tts.py:197]: Aren't these batches?

File: deeppavlov/models/nemo/tts.py [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+        transcript_embedded = self.text_embedding(char_phone=transcript)
+        transcript_encoded = self.t2_enc(char_phone_embeddings=transcript_embedded, embedding_length=transcript_len)
+        mel_decoder, gate, alignments, mel_len = self.t2_dec(char_phone_encoded=transcript_encoded,
+                                                             encoded_length=transcript_len)
+        mel_postnet = self.t2_postnet(mel_input=mel_decoder)
+        infer_tensors = [self.vocoder(mel_postnet), mel_len]
+        evaluated_tensors = self.neural_factory.infer(tensors=infer_tensors)
+        synthesized_batch = self.vocoder.get_audio(*evaluated_tensors)
+
+        for fout, data in zip(path_batch, synthesized_batch):
```
Comment (oserikov) [inline deeppavlov/models/nemo/tts.py:207]: Is data unambigous here?

File: deeppavlov/models/nemo/vocoder.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+        raise NotImplementedError
+
+
+class WaveGlow(BaseVocoder):
+    def __init__(self, *, denoiser_strength: float = 0.0, n_window_stride: int = 160, **kwargs) -> None:
+        """Wraps WaveGlowInferNM module.
+
+        Args:
+            denoiser_strength: Denoiser strength for waveglow.
+            n_window_stride: Stride of window for fft in samples used in model training.
```
Comment (oserikov) [inline deeppavlov/models/nemo/vocoder.py:45]: fft -> FFT ?

File: docs/features/models/nemo.rst [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+any additional alignment information. :class:`~deeppavlov.models.nemo.asr.NeMoASR` contains following modules:
+
+-  `AudioToMelSpectrogramPreprocessor <https://github.com/NVIDIA/NeMo/blob/v0.10.0/nemo/collections/asr/audio_preprocessing.py>`_ - uses arguments from ``AudioToMelSpectrogramPreprocessor`` section of the NeMo config file.
+-  `JasperEncoder <https://nvidia.github.io/NeMo/collections/nemo_asr.html#nemo.collections.asr.jasper.JasperEncoder>`__ - uses arguments from ``JasperEncoder`` section of the NeMo config file. Needs pretrained checkpoint.
+-  `JasperDecoderForCTC <https://nvidia.github.io/NeMo/collections/nemo_asr.html#nemo.collections.asr.jasper.JasperDecoderForCTC>`__ - uses arguments from ``JasperDecoder`` section of the NeMo config file. Needs pretrained checkpoint.
+-  `GreedyCTCDecoder <https://github.com/NVIDIA/NeMo/blob/v0.10.0/nemo/collections/asr/greedy_ctc_decoder.py>`__ - doesn't use any arguments.
+-  :class:`~deeppavlov.models.nemo.asr.AudioInferDataLayer` - uses arguments from ``AudioToTextDataLayer`` section of the NeMo config file.
+
+Besides named arguments for the modules above NeMo config file for ASR should contain ``labels`` arg.
```
Comment (oserikov) [inline docs/features/models/nemo.rst:21]: Probably we could clarify what labels arg stands for

File: docs/features/models/nemo.rst [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+-  `Tacotron2DecoderInfer <https://nvidia.github.io/NeMo/collections/nemo_tts.html#nemo.collections.tts.tacotron2_modules.Tacotron2Decoder>`__ - uses arguments from ``Tacotron2Decoder`` section of the NeMo config file. Needs pretrained checkpoint.
+-  `Tacotron2Postnet <https://nvidia.github.io/NeMo/collections/nemo_tts.html#nemo.collections.tts.tacotron2_modules.Tacotron2Postnet>`__ - uses arguments from ``Tacotron2Postnet`` section of the NeMo config file. Needs pretrained checkpoint.
+-  :class:`~deeppavlov.models.nemo.vocoder.WaveGlow` - uses arguments from ``WaveGlowNM`` section of the NeMo config file. Needs pretrained checkpoint.
+-  :class:`~deeppavlov.models.nemo.vocoder.GriffinLim` - uses arguments from ``GriffinLim`` section of the NeMo config file.
+-  :class:`~deeppavlov.models.nemo.tts.TextDataLayer` - uses arguments from ``TranscriptDataLayer`` section of the NeMo config file.
+
+NeMo config file for ASR should contain ``labels`` and ``sample_rate`` args besides named arguments for the modules above.
```
Comment (oserikov) [inline docs/features/models/nemo.rst:37]: Probably we could clarify what labels and sample rate args do stand for

File: deeppavlov/models/nemo/asr.py (hunk 4 above)
Comment (IgnatovFedor) [reply deeppavlov/models/nemo/asr.py:43]: I decided to use the same argument name that uses a similar class to maintain uniformity: ``nemo.collections.asr.parts.dataset.AudioDataset`` uses ``trim`` argument name, as well as ``nemo.collections.asr.parts.features.WaveformFeaturizer.process`` method.

File: deeppavlov/models/nemo/asr.py (hunk 6 above)
Comment (IgnatovFedor) [reply deeppavlov/models/nemo/asr.py:190]: What variable name would you use? Rename `eval_tensors` to `tensors` and `tensors` to `evaluated_tensors`?

File: deeppavlov/models/nemo/asr.py (hunk 7 above)
Comment (IgnatovFedor) [reply deeppavlov/models/nemo/asr.py:191]: In my opinion it's excessive: In that case, I should rename `audio_batch` to `audio_batch_to_transcribe` or something. Current naming is quite concise and clear: method receives `audio_batch`, returns `text_batch`.

File: deeppavlov/models/nemo/common.py (hunk 10 above)
Comment (IgnatovFedor) [reply deeppavlov/models/nemo/common.py:82]: I agree, `device` or `device_type` are better names, but I decided to use the same variable name as `NeuralModuleFactory.__init__` argument name.

File: deeppavlov/models/nemo/common.py (hunk 12 above)
Comment (IgnatovFedor) [reply deeppavlov/models/nemo/common.py:98]: Yes, it is: parent class `Serializable` has abstract method `save`.

File: deeppavlov/models/nemo/tts.py (hunk 13 above)
Comment (IgnatovFedor) [reply deeppavlov/models/nemo/tts.py:124]: "not unambigous" == "ambigous"?) Got it `s` -> `text`

File: deeppavlov/models/nemo/tts.py (hunk 14 above)
Comment (IgnatovFedor) [reply deeppavlov/models/nemo/tts.py:197]: Yes and no, this is NmTensors

File: docs/features/models/nemo.rst (hunk 18 above)
Comment (IgnatovFedor) [reply docs/features/models/nemo.rst]: Labels description has added, but sample rate speaks for itself, so I didn't add it.
